src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,639 @@
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.Iterator;
+import java.util.Objects;
+
+/**
+ * This class provides the functionality of a Key Derivation Function (KDF),
+ * which is a cryptographic algorithm for deriving additional keys from input
+ * keying material and (optionally) other data.
+ * <p>
+ * {@code KDF} objects are instantiated with the {@code getInstance} family
+ * of methods. For HKDF, KDF algorithm names follow a naming convention of
+ * <em>Algorithm</em>With<em>PRF</em>. For instance, a KDF implementation of
+ * HKDF using HMAC-SHA256 has an algorithm name of "HKDFWithHmacSHA256". For
+ * future algorithms, the "WithPRF" portion of the algorithm field may be
+ * omitted if the KDF algorithm has a fixed or default PRF or if the PRF is
+ * specified as part of the {@code KDFParameters}.
+ * <p>
+ * The class has two derive methods, {@code deriveKey} and {@code deriveData}.
+ * The {@code deriveKey} method accepts an algorithm {@code String} and
+ * will return a {@code SecretKey} object with the specified algorithm. The
+ * {@code deriveData} method returns a byte array of raw data.
+ * <p>
+ * If a provider is not specified in the {@code getInstance} method when
+ * instantiating a {@code KDF} object, the provider is selected the first time
+ * the {@code deriveKey} or {@code deriveData} method is called, and a provider
+ * is chosen that supports the parameters passed to the {@code deriveKey} or
+ * {@code deriveData} method. However, if {@code getProviderName} or
+ * {@code getParameters} is called before calling the {@code deriveKey} or
+ * {@code deriveData} methods, the first provider supporting the KDF algorithm
+ * and {@code KDFParameters} is chosen which may not be the provider that is
+ * eventually selected once the {@code AlgorithmParameterSpec} is supplied in
+ * the derive methods; therefore it is recommended not to call
+ * {@code getProviderName} or {@code getKDFParameters} until after a key
+ * derivation operation. Once a provider is selected, it cannot be changed.
+ * <p>
+ * API Usage Example:
+ * {@snippet lang = java:
+ *    KDF kdfHkdf = KDF.getInstance("HKDFWithHmacSHA256");
+ *
+ *    AlgorithmParameterSpec derivationParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *    kdfHkdf.deriveKey("AES", derivationParameterSpec);
+ *}
+ *
+ * @see KDFParameters
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KDF");
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
```

Comment:
Add some debug output, especially on exceptions for all different reasons during delayed provider selection.